nvflare/private/fed/server/server_engine.py

```diff
@@ -370,13 +370,18 @@ def abort_app_on_server(self, job_id: str) -> str:
                 if child_process:
                     child_process.terminate()
         finally:
-            with self.lock:
-                if job_id in self.run_processes:
-                    self.run_processes.pop(job_id)
+            threading.Thread(target=self._remove_run_processes, args=[job_id]).start()
 
         self.engine_info.status = MachineStatus.STOPPED
         return ""
 
+    def _remove_run_processes(self, job_id):
+        # wait for the run process to gracefully terminated, and ensure to remove from run_processes.
+        time.sleep(60.0)
```

Comment:
Note that this does not guarantee the run process has gracefully terminated though.

We just sleep 60 seconds and assume that.

Reply:
correct. The abort command is sent. We can just wait 60 seconds assuming it shutdown self. Then terminate it any way.